[#50] 주문 결제 기능 구현

chicori3 opened the pull request:
## 목표

유저는 생성된 주문을 결제할 수 있다

## 할 일

- [ ] 결제 모듈 설정 - 추후 고민
- [x] 결제 테스트 작성
- [x] 결제 성공 시 주문 상태 변경
- [x] 결제 실패 시 주문 상태와 쿠폰 상태 변경

## 고려사항

- 결제 모듈 같은 외부 API를 통합하는 것은 멘토링 시간에 미루기로 결정했습니다.
- 결제 시 주문의 상태 또는 결제 금액이 일치하지 않은 경우 예외가 발생하는데, 트랜잭션은 유지하여 주문과 쿠폰 상태를 변경할 수 있도록 Transactional의 `dontRollbackOn` 옵션을 사용했습니다. 이 방법말고 다른 방법이 더 있을까요?

chicori3 committed:
refactor(#6) - Coupon 도메인 수정, 에러 코드 추가, 메서드 참조 식으로 변경

chicori3 committed:
feat(#6) - Order 도메인 작성

chicori3 committed:
feat(#6) - BaseEntity 작성, JpaAuditing 설정

chicori3 committed:
feat(#6) - 주문 생성 리포지토리 코드 작성

chicori3 committed:
feat(#6) - 주문 생성 서비스 계층 작성

chicori3 committed:
refactor(#6) - 쿠폰 리포지토리 Stub 네이밍 변경, DiscountType 할인값을 BigDecimal 타입으로 변경

chicori3 committed:
refactor(#6) - 사용하지 않는 CouponProcessor 삭제

chicori3 committed:
feat(#6) - 장바구니에 해당 강의가 있는지 검증하는 로직 추가

chicori3 committed:
feat(#6) - 회원 Role을 Teacher로 승급하는 로직 추가

chicori3 committed:
feat(#6) - 테스트를 위한 Fixture 작성

chicori3 committed:
refactor(#6) - 도메인 로직 수정, OrderItem 엔티티 작성, 에러 코드 추가

chicori3 committed:
refactor(#6) - 주문 생성 로직 수정

chicori3 committed:
feat(#6) - 통합 테스트 기본 클래스와 DB 초기화 객체 작성

chicori3 committed:
test(#6) - 주문 생성 테스트 작성

chicori3 committed:
chore(#6) - application.yaml 수정

chicori3 committed:
feat(#6) - 주문 생성 컨트롤러 작성

chicori3 committed:
test(#6) - 주문 생성 컨트롤러 테스트 작성

chicori3 committed:
refactor(#6) - 주문한 강의 정보 추가

chicori3 committed:
refactor(#6) - Cart 도메인 수정

- CartItem 엔티티 승격
- Cart-CartItems 일대다 연관관계 설정
- Fixture 수정

chicori3 committed:
refactor(#6) - Cart 리포지토리 수정

- CartItem 까지 조회하는 join fetch 쿼리 추가

chicori3 committed:
refactor(#6) - Cart 서비스 계층 수정

- Cart 도메인 수정으로 인한 변경

chicori3 committed:
refactor(#6) - 피드백 반영

- 부적절한 메서드 네이밍 수정
- DiscountValue의 타입을 BigDeicmal -> int로 변경
- 주문 생성 시 쿠폰을 확인하고 반환할 때 Optional 타입으로 반환

chicori3 committed:
refactor(#6) - DatabaseCleaner가 작동하도록 @Table 어노테이션 추가

chicori3 committed:
test(#6) - 주문 생성 테스트 케이스 추가 작성

chicori3 committed:
feat(#50) - .

chicori3 committed:
feat(#50) - Payment 도메인 작성

chicori3 committed:
feat(#50) - Order 결제 로직 작성, 예외 추가

chicori3 committed:
feat(#50) - Payment 리포지토리 작성

chicori3 committed:
feat(#50) - 주문 결제 서비스 작성

chicori3 committed:
refactor(#50) - 도메인 코드 수정

- Order 도메인 생성 시 가격 값을 받아 Price를 생성하도록 변경
- Member 엔티티의 값이 필요하지 않아 Member 엔티티 참조 대신 ID 참조로 변경
- Price 비교 메서드 추가

chicori3 committed:
refactor(#50) - 도메인 코드 수정으로 인한 픽스처 수정

chicori3 committed:
refactor(#50) - Price 값 객체 생성 책임을 Order 도메인의 책임으로 변경

chicori3 committed:
test(#50) - 주문 결제 서비스 테스트 작성

chicori3 committed:
feat(#50) - 주문 결제 컨트롤러 작성

chicori3 committed:
test(#50) - 주문 컨트롤러 주문 결제 테스트 작성

chicori3 committed:
feat(#50) - 주문 취소 로직 작성

- 결제 실패 시 주문 상태를 주문 취소로 변경
- 쿠폰을 미사용 상태로 변경
- 쿠폰의 상태를 변경하기 위해 로딩 전략을 EAGER로 수정
- 예외가 발생해도 트랜잭션을 롤백시키지 않기 위해 dontRollbackOn 옵션 사용

chicori3 committed:
refactor(#50) - 결제 실패 테스트 수정

- 결제 실패 시 예외와 주문 상태, 쿠폰 상태를 검증

kyeumd commented on src/main/java/com/flab/infrun/cart/infrastructure/persistence/CartJpaRepository.java:
```diff
@@ -3,8 +3,13 @@
 
     Optional<Cart> findByOwnerId(final Long ownerId);
+
+    @Query("select distinct c from Cart c join fetch c.cartItems.cartItems where c.ownerId = :ownerId")
```
distinct를 사용하신 이유가 있을까요?

chicori3 replied:
@OneToMany 연관관계에서 fetch join을 사용하면 쿼리의 결과가 카테시안 곱으로 나올 수 있기 때문에 distinct를 사용했습니다

kyeumd commented on src/main/java/com/flab/infrun/coupon/infrastructure/CouponValidatorImpl.java:
```diff
@@ -18,7 +18,7 @@ public void verifyDiscountType(final String discountType) {
 
     @Override
-    public void verifyDiscountAmount(final int discountAmount) {
+    public void verifyDiscountValue(final int discountAmount) {
```
Class 명의 Validator와 문맥이 비슷한 validate 용어 대신 verify를 사용하신 이유가 있을까요?

chicori3 replied:
작성하면서 verify와 validate의 차이에 대해서 고민했습니다

validate는 **사용자의 입력을 검증**하는 역할이라고 생각했고 실제로 Request 객체의 필드가 유효한지 spring-boot-starter-validation 라이브러리의 @Valid 어노테이션을 사용합니다

verify는 도메인 로직, 코드 레벨에서의 과정을 검증하는 것이라고 생각합니다.

kyeumd commented on src/main/java/com/flab/infrun/coupon/domain/Coupon.java:
```diff
@@ -65,6 +70,47 @@ public static Coupon create(
             .build();
     }
 
+    public void enroll(final Member owner, final LocalDateTime currentTime) {
```
coupon의 **enroll 메서드**와  
current time으로 coupon의 status를 관리하는 등록가능 여부를 체크하는 것으로 보이는 **verifyIsRegistrable 메서드**가 어떤 의도를 위해 작성된 걸까요?

chicori3 replied:
질문의 의도를 잘 파악하지 못했습니다.

enroll 메서드는 쿠폰에 유저를 등록시키는 메서드이고,
verifyIsRegistrable은 유저를 등록시키기 전에 등록이 가능한 상태인지 검증하는 메서드입니다.

kyeumd reviewed (commented):
(아마 멘토님의 리뷰가 즉각적으로 이루어지기 힘들기 때문에 그런 것으로 생각되긴 합니다만)
현재 PR의 commit 내역이 다소 많아서 의미있는 단위로 파악을 하기가 어려운 것 같습니다.
common 부분의 변경사항은 따로 PR을 작성하는 것이 좋을 것 같고
commit이 많아지는 경우 PR 단위 자체를 분리하여 작성하는 것은 어떨까요?

f-lab-michael commented on src/main/java/com/flab/infrun/coupon/domain/Coupon.java:
```diff
@@ -65,6 +70,47 @@ public static Coupon create(
+    }
+
+    private void verifyIsRegistrable(final LocalDateTime currentTime) {
+        if (this.status != CouponStatus.UNREGISTERED) {
+            throw new AlreadyRegisteredCouponException(ErrorCode.ALREADY_REGISTERED_COUPON);
+        }
+        if (this.expirationAt.isBefore(currentTime)) {
+            throw new ExpiredCouponException(ErrorCode.EXPIRED_COUPON);
+        }
```
상태에 EXPIRED 가 있는데 해당부분을 먼저 체크해서 만료 여부체크후에 이미 사용중인지 등록되었는지등을 체크해야 하지 않을까요?

chicori3 replied:
이 부분은 미처 생각못한 부분입니다 감사합니다

f-lab-michael commented on src/main/java/com/flab/infrun/coupon/domain/Coupon.java:
```diff
@@ -65,6 +70,47 @@ public static Coupon create(
+            throw new InvalidCouponOwnerException();
+        }
+        if (this.expirationAt.isBefore(currentTime)) {
+            this.status = CouponStatus.EXPIRED;
```
verify 라고 이름지었기 때문에 검증만 하는것이 좋습니다. 만약 쿠폰의 상태를 변경하고 싶다면 verifyIsUsableAndExpireIfneeded 같이 메소드 이름에 만료도 시킬수있다는 사실을 표현하는게 좋습니다.

chicori3 replied:
@f-lab-michael 오 한가지 생각난 것은 매일 00시에 쿠폰의 상태를 일괄 변경하는 기능이 있으면 좋을 것 같습니다.
Batch에 대해 알아보면 될까요??

f-lab-michael replied:
별도 배치 프로세스를 도입하는것보다는 간단하게 @Scheduled 어노테이션좀 살펴보시고 적용하면 될것 같아요. 
대신 어플리케이션 서버가 여러대가 되었을때 어떻게 될지에 대한 고민은 필요해보입니다.

f-lab-michael commented on src/main/java/com/flab/infrun/order/presentation/request/PayOrderRequest.java:
```diff
@@ -0,0 +1,38 @@
+    @NotNull(message = "결제 금액은 필수입니다.")
+    BigDecimal amount,
+    @NotBlank(message = "결제 방법은 필수입니다.")
+    @Pattern(regexp = "\\bCARD\\b|\\bCASH\\b", message = "결제 방법은 CARD 또는 CASH만 가능합니다.")
+    String payMethod,
```
제약을 걸꺼면 enum으로 처리 하는건 어떤가요?

f-lab-michael commented on src/main/java/com/flab/infrun/order/presentation/request/PayOrderRequest.java:
```diff
@@ -0,0 +1,38 @@
+    String payMethod,
+
+    @NotBlank(message = "결제 타입은 필수입니다.")
+    @Pattern(regexp = "\\bLUMP_SUM\\b|\\bINSTALLMENT\\b", message = "결제 타입은 LUMP_SUM 또는 INSTALLMENT만 가능합니다.")
+    String payType
```
여기도 제약을 걸꺼면 enum으로 처리하는건 어떤가요?

f-lab-michael commented on src/main/java/com/flab/infrun/payment/domain/Payment.java:
```diff
@@ -0,0 +1,70 @@
+    private PayType payType;
+    @Enumerated(EnumType.STRING)
+    private PayMethod payMethod;
+    private boolean isSuccess;
+    private boolean isCanceled;
```
여러가지 상태가 존재 하고 하나의 상태만 가질수있다면 별도로 상태값을 정의해서 쓰는건 어떨까요?

chicori3 replied:
이것도 생각못했네요 감사합니다!

f-lab-michael reviewed (commented):
변경되는 사항에 비해서 PR크기가 좀 큰편인것 같습니다. 
동겸님 말처럼 하나의 PR에는 가능하면 하나의 주요 변경사항만 들어가는것이 좋습니다. 
가능하면 부가적으로 추가하는것들에 대해서는 PR을 쪼개는게 좋습니다.

chicori3 committed:
refactor(#3) - 피드백 반영

- 요청 객체 필드의 Enum 타입을 검증하는 EnumValidator 작성 및 적용
- PayStatus를 추가하여 하나의 상태로 반영하도록 수정
- Coupon의 등록, 사용 검증 로직에서 상태를 변경하는 것을 명시하도록 메서드 네이밍 수정

chicori3 committed:
Merge branch 'feature/#6-order-service' into feature/#50-order-pay
